### ISSUE TITLE
[otbn, sw] assumptions of barrett384.s

### ISSUE THREAD

**Issue by yizhou7** [lowRISC/opentitan#4548]
Following up on a discussion with @felixmiller : 

Using the terms from 14.42 in "Handbook of Applied Cryptography",  `b` is the radix, and `k` is some exponent. The book version assumes a general `m`, so the bound for `q3` is `Q ≤ q3 ≤ Q + 2`. The book assumes that the radix `b > 3`, so the inequality `(Q - q3) * m + R < 3 * m < b ^ (k + 1)` holds.

The current implementation of `barrett384` is using radix `b = 2`, and exponent `k = 384`. If we assume generally that `2^383 < m < 2^384`, then the inequality should not hold. 

However, as @felixmiller pointed out earlier, using the modulus of p384 and p256, it should be possible to show that `Q ≤ q3 ≤ Q + 1`. Then we can show that `(Q - q3) * m + R < 2 * m < b ^ (k + 1)` holds. In that case it might be possible to remove one conditional subtraction as well. 

To summarize, the current barrett384.s: 
* should have this tighter bound for `q3` when working with particular modulus, and might have optimization potential
* might need a fix for general moduli. 

Our understanding is that there will be different versions for each, and we are interested in seeing them when available. 

> estimate 8
> remaining 2021-01-25 1

**Comment by felixmiller** [comment]
Reworked the code now for the generic version as part of [this](https://github.com/lowRISC/opentitan/pull/4957) PR. In the new version we do not do the truncation of step 2 of HAC 14.42 and use the full register width. i.e. we use 512 bit for r2 and r. r1 is just the two lower limbs of x then. We don't really benefit from any truncation, since we're using radix 2 but have 256 bit wide registers. 
This also allows us to remove:
- The special treatment of negative r (r cannot be negative now)
- The computation of the mask that was used for the truncation
- Computing r1 at the beginning just after computation of x

Overall these are less instructions than in the full HAC version, therefore we plan to use this approach also in the P384 optimized version (which will arrive soon).

**Comment by imphil** [comment]
Fixed by https://github.com/lowRISC/opentitan/pull/4957
